[RDY] Improve functional test debuggability and efficiency

tarruda opened the pull request:
- Read TEST_TAG env var from cmake/RunTests.cmake. Setting this environment
  variable will pass --tags with its value to busted, which is used to filter
  which tests will be executed.
- Remove calls to nvim msgpack-rpc API outside tests. This removes the
  requirement of having a static `clear` call in test/functional/helpers.lua
- Replace all `setup` calls by the `before_first` function  which is injected on
  the top of each test file that needs to have suite setup. This differs from
  `setup` in that the callback will only be executed if at least one test will
  be ran in the suite, where `setup` always executes the callback regardless of
  test filters.

After this commit its possible to run/debug a single test like this:

``` sh
TEST_TAG=some-tag make test
```

After putting #some-tag into the test title. This also improves debugging
experience because there will be no unnecessary gdbserver instances whe GDB=1 is
passed.

justinmk commented on test/functional/clipboard/clipboard_provider_spec.lua:
```diff
@@ -62,7 +63,7 @@ local function basic_register_test()
 
 describe('clipboard usage', function()
-  setup(reset)
+  before_first(reset)
```
How is this different than `setup()`, which only runs once per `describe()` block?

justinmk replied:
Oops, I just read the commit message.

bfredl commented on test/functional/ui/screen.lua:
```diff
@@ -496,8 +501,8 @@ function pprint_attrs(attrs)
-        if colornames[v] ~= nil then
-          desc = "Screen.colors."..colornames[v]
+        if Screen._colornames[v] ~= nil then
+          desc = "Screen.colors."..Screen._colornames[v]
```
`Screen.colors().` for consistency here, but tell me, why should `Screen.colors` be a function? It is static data that doesn't depend on the state of nvim.

justinmk commented:
> This differs from  `setup` in that the callback will only be executed if at least one test will  be ran in the suite, where `setup` always executes the callback regardless of  test filters.

I assume this only matters if there is more than one `describe()` block in a given test file. Otherwise, couldn't you specify:

    TEST_FILE=foo TEST_TAG=some-tag make test

Is it correct that `before_first()` only saves a couple unnecessary `clear()` calls for a given test file? Does an extra `clear()` call complicate debugging?

bfredl commented:
If it turns out we want `before_first()`, why not define it in `helpers.lua` instead of replicating it in every test file? Or even just `before_each(once(clear)))` at the call site?

tarruda commented:
@bfredl true, but it still requires spawning a nvim instance with `clear`. Though for static data it might be better to spawn nvim directly with the lua client API

@justinmk the goal is not have to specify a test file, but only a tag and let busted do the filtering. 

Extra calls to `clear` complicate debugging because each call before the test we need to debug means we have to attach to gdbserver one time, run until completion and then reattach. This is especially boring when running nvim with valgrind's gdbserver since the startup time is really long, and even more when setting breakpoints that might be hit even before we get to the desired test.

bfredl commented:
and BTW, unrelated to this PR, but, can't we just the moral equivalent of `from helpers.lua import *` ? Yes I know explicit imports are "better" but frankly, do those `local clear, nvim, buffer, curbuf, curwin, eq, neq, ok =
helpers.clear, helpers.nvim, helpers.buffer, helpers.curbuf, helpers.curwin,
helpers.eq, helpers.neq, helpers.ok` serve any good? (And the busted defines are implicit anyway...)

tarruda commented:
@bfredl because `before_each` is not available globally, only in the test environment. It would be useful if we could force a test file to be ran before others so we could set it as a global function

bfredl commented:
@tarruda True, meant it mostly from the API design standpoint, `Screen.colors.name` seems more logical. (Even an attribute might be generated on first use)

justinmk commented:
@bfredl I assume busted is modifying `_G` or something; we could selectively (judiciously...) do something similar. But http://www.lua.org/notes/ltn007.html is a good practice to avoid time-sinking scope bugs.

bfredl commented:
Alternatively, `colors` could be an attribute of a `screen` _instance_ (and calculated in `Screen.new`) that would makes the dependency on an actual nvim instance explicit.

tarruda commented:
@justinmk Something like that , but I dont think we have the option of overriding busted definitions globally

@bfredl lua supports overriding table indexing, so Screen.colors can be a method with field syntax

I'm gonna hold on this PR for a week to see if busted#377 is fixed

bfredl commented:
@justinmk True, but I don't think an individual test file as a "module" strictly, as as you pointed out, it already runs in a busted-defined envirionment, and `helpers.lua` could be considered to be an extension of  the busted "language". 
But my main point is that, is really `local clear, nvim, buffer, curbuf, curwin, eq, neq, ok = helpers.clear, helpers.nvim, helpers.buffer, helpers.curbuf, helpers.curwin, helpers.eq, helpers.neq, helpers.ok` any better? What if LHS and RHS gets out of sync, now that would be a potential source of very confusing errors

bfredl commented:
And I mean, not neccessarily to try to change the predefined envirionment in the background, rather there could be something like `helpers.importall()`.

tarruda commented:
@bfredl :+1: yes, removing all those imports is a good idea

Oberon00 commented on test/functional/helpers.lua:
```diff
@@ -271,7 +271,12 @@ local function expect(contents)
 end
 
-clear()
+function once(fn)
```
Wouldn't an `assert(fn)` make sense here? Cause when you accidentally pass `nil` (which happens quite easy in Lua) the `if` below will lead to silently doing nothing.

tarruda replied:
:+1:

tarruda committed:
deps: Update busted and dependencies

tarruda committed:
test: Extract code to spawn nvim into the "spawn" helper function

This is can be used for spawning nvim outside a test context. Also refactor
screen.lua to use this function when loading the color map(It is better because
the GDB/VALGRIND environment variables are ignored)

tarruda committed:
test: Improve functional test debuggability and efficiency

- Read TEST_TAG/TEST_FILTER env vars from cmake/RunTests.cmake. Setting these
  environment variables will pass --tags/--filter to busted, which can used to
  filter which tests are executed.
- Remove calls to nvim msgpack-rpc API outside tests. This removes the
  requirement of having a static `clear` call in test/functional/helpers.lua
- Use the new busted command-line option "--lazy" to ensure the setup/teardown
  hooks are only executed when a suite runs at least one test.

Now its possible to run/debug a single test like this:

```sh
TEST_FILTER='some test string' make test
```

Which will only run tests containing "some test string" in the title.
Another option is:

```sh
TEST_TAG=some-tag make test
```

After putting #some-tag into the test title. This also improves debugging
experience because there will be no unnecessary gdbserver instances whe GDB=1 is
passed.

tarruda commented:
Removed the `before_first`/`once` hack since busted now supports the `--lazy` option to achieve the same behavior. Thanks @o-lim for implementing this!

bfredl commented on test/functional/helpers.lua:
```diff
@@ -162,15 +162,20 @@ local function rawfeed(...)
   end
 end
 
+local function spawn(argv)
```
Nice! I will update #2077 after this PR is done.

tarruda replied:
:+1:

tarruda commented:
I'm merging this after the tests. I was going to wait for busted to make a release but its taking too long(they said a release would be made 7 days ago). When they decide to release a new version we'll simply update it.

justinmk commented:
> the goal is not have to specify a test file, but only a tag and let busted do the filtering.

Does the busted `--filter` flag do what you want here? See https://github.com/neovim/neovim/pull/2851#issuecomment-114060470

tarruda commented:
> @tarruda Does the busted --filter flag do what you want here? See #2851 (comment)

Yes, that's why I added the `TEST_FILTER` environment variable(to be able to use `--filter` while still abstracting the test run through `make test`)
